xlsx format breaks retrieval function

Trying to run retrieval function on [Brazil Minas Gerais dataset](coronavirus.saude.mg.gov.br/images/microdados/xlsx_sistemas.xlsx), which is in .xlsx format. Retrieval function breaks with following error message:

```
$ sam local invoke "RetrievalFunction" -e retrieval/valid_scheduled_event.json --docker-network=host
Invoking retrieval.lambda_handler (python3.8)
CommonLibLayer is a local Layer in the template
Building image.................
Requested to skip pulling images ...

Mounting /Users/felson/Documents/DPhil/global_health/github/list/ingestion/functions/.aws-sam/build/RetrievalFunction as /var/task:ro,delegated inside runtime container
START RequestId: d3fcd181-71fc-1432-9d0f-3b671bd6ec79 Version: $LATEST
Extracting fields from event {'env': 'local', 'sourceId': '5f3ce726c7416b00345302f8', 'parsingDateRange': {'start': '2020-09-01', 'end': '2020-09-21'}}
Retrieving service account credentials from s3://epid-ingestion/covid-19-map-277002-0943eeb6776b.json
Creating upload via http://localhost:3001/api/sources/5f3ce726c7416b00345302f8/uploads
Requesting source configuration from http://localhost:3001/api/sources/5f3ce726c7416b00345302f8
Received source API response: {'notificationRecipients': [], '_id': '5f3ce726c7416b00345302f8', 'name': 'minas_gerais_test', 'origin': {'url': 'coronavirus.saude.mg.gov.br/images/microdados/xlsx_sistemas.xlsx'}, 'format': 'XLSX', 'uploads': [{'created': '2020-09-02T12:05:07.326Z', '_id': '5f4f8a739767ff004b750975', 'status': 'IN_PROGRESS', 'summary': {}}, {'created': '2020-09-30T10:29:26.753Z', '_id': '5f745e0625c9ba00397dbc65', 'status': 'ERROR', 'summary': {'error': 'INTERNAL_ERROR'}}, {'created': '2020-09-30T10:34:45.748Z', '_id': '5f745f4525c9ba00397dbc66', 'status': 'ERROR', 'summary': {'error': 'INTERNAL_ERROR'}}, {'created': '2020-09-30T10:46:18.068Z', '_id': '5f7461fa25c9ba00397dbc67', 'status': 'ERROR', 'summary': {'error': 'INTERNAL_ERROR'}}, {'created': '2020-12-03T07:21:14.357Z', '_id': '5fc891ea345a415138c84a54', 'status': 'ERROR', 'summary': {'error': 'SOURCE_CONFIGURATION_ERROR'}}, {'created': '2020-12-16T18:30:03.769Z', '_id': '5fda522b168a0f007f848a7b', 'status': 'ERROR', 'summary': {'error': 'SOURCE_CONFIGURATION_ERROR'}}, {'created': '2020-12-18T11:24:28.729Z', '_id': '5fdc916c367aca00304f9952', 'status': 'IN_PROGRESS', 'summary': {}}], '__v': 8}
Unsupported source format: XLSX
Updating upload via http://localhost:3001/api/sources/5f3ce726c7416b00345302f8/uploads/5fdc916c367aca00304f9952
[ERROR] ValueError: Unsupported source format: XLSX
Traceback (most recent call last):
  File "/var/task/retrieval.py", line 265, in lambda_handler
    file_name, s3_object_key = retrieve_content(
  File "/var/task/retrieval.py", line 114, in retrieve_content
    common_lib.complete_with_error(
  File "/opt/python/common_lib.py", line 98, in complete_with_error
    raise exception
END RequestId: d3fcd181-71fc-1432-9d0f-3b671bd6ec79
REPORT RequestId: d3fcd181-71fc-1432-9d0f-3b671bd6ec79  Init Duration: 3564.61 ms       Duration: 1625.77 ms    Billed Duration: 1700 ms        Memory Size: 1024 MB    Max Memory Used: 50 MB

{"errorType":"ValueError","errorMessage":"Unsupported source format: XLSX","stackTrace":["  File \"/var/task/retrieval.py\", line 265, in lambda_handler\n    file_name, s3_object_key = retrieve_content(\n","  File \"/var/task/retrieval.py\", line 114, in retrieve_content\n    common_lib.complete_with_error(\n","  File \"/opt/python/common_lib.py\", line 98, in complete_with_error\n    raise exception\n"]}
```

--- a/ingestion/functions/retrieval/retrieval.py
+++ b/ingestion/functions/retrieval/retrieval.py
@@ -2,7 +2,6 @@ import codecs
 import io
 import mimetypes
 import os
-import re
 import sys
 import tempfile
 import zipfile
@@ -145,6 +144,11 @@ def retrieve_content(
         print('Download finished')
 
         key_filename_part = f"content.{source_format.lower()}"
+        s3_object_key = (
+            f"{source_id}"
+            f"{datetime.now(timezone.utc).strftime(TIME_FILEPART_FORMAT)}"
+            f"{key_filename_part}"
+        )
         # Lambda limitations: 512MB ephemeral disk space.
         # Memory range is from 128 to 3008 MB so we could switch to
         # https://docs.python.org/3/library/io.html#io.StringIO for bigger
@@ -152,7 +156,16 @@ def retrieve_content(
         # Make the encoding of retrieved content consistent (UTF-8) for all
         # parsers as per https://github.com/globaldothealth/list/issues/867.
         bytesio = raw_content(url, content, tempdir)
-        print('detecting encoding of retrieved content.')
+        if source_format == "XLSX":
+            # do not convert XLSX into another encoding, leave for parsers
+            print("Skipping encoding detection for XLSX")
+            fd, outfile_name = tempfile.mkstemp(dir=tempdir)
+            with os.fdopen(fd, "wb") as outfile:
+                while content := bytesio.read(READ_CHUNK_BYTES):
+                    outfile.write(content)
+            return [(outfile_name, s3_object_key)]
+
+        print('Detecting encoding of retrieved content')
         # Read 2MB to be quite sure about the encoding.
         detected_enc = detect(bytesio.read(2 << 20))
         bytesio.seek(0)
@@ -170,10 +183,6 @@ def retrieve_content(
             while content:
                 outfile.write(content)
                 content = text_stream.read(READ_CHUNK_BYTES)
-            s3_object_key = (
-                f"{source_id}"
-                f"{datetime.now(timezone.utc).strftime(TIME_FILEPART_FORMAT)}"
-                f"{key_filename_part}")
             return [(outfile_name, s3_object_key)]
     except requests.exceptions.RequestException as e:
         upload_error = (

